New Models Implementation

nikit-srivastava opened the pull request:
Following is the description of classes implemented:
#### 1. DataSubsetGenerator.java : 
Class to help generate and persist the subsets of a word2vec model
#### 2. DataSubsetProvider.java : 
Class to help retrieve the list of words from a subset stored on the disk
#### 3. ModelToCSVWriter.java : 
Class to help write a word2vec model to a csv file

Additionally a new method is introduced:

#### 4. Word2VecModel.getClosestEntryInSub(float[] inpvec, String subsetKey) : 
Method to fetch a single closest word entry to the passed input vector inside the subset of word2vec model

nikit-srivastava committed:
Subset logic implemented

nikit-srivastava committed:
javadoc comments + makedir

nikit-srivastava committed:
cleaning pom of extra entries

nikit91 commented:
@MichaelRoeder please review

MichaelRoeder commented on src/main/java/org/aksw/word2vecrestful/subset/DataSubsetProvider.java:
```diff
@@ -0,0 +1,70 @@
+ * @author Nikit
+ *
+ */
+public class DataSubsetProvider {
```
I am not sure why you are creataing JSON files when you want to write a list of words. Why not simply writing one word per line? There are FileUtils methods that can do that very easily ;)

MichaelRoeder commented on src/main/java/org/aksw/word2vecrestful/word2vec/Word2VecModel.java:
```diff
@@ -74,6 +79,36 @@ private void fetchVectorDimensions() {
+	  float[] minVec = null;
+	  double tempDist;
+	  // Loop on the subset
+	  for(String word : DataSubsetProvider.fetchSubsetWords(subsetKey)) {
```
This looks like you have to load the subset from file for every single request. Wouldn't it be better to have an instance of the DataSubsetProvider class that is able to simply store the loaded list in memory? As we already discussed, this wouldn't consume a lot of memory since the Strings are already loaded in the JVM because of the in-memory Word2Vec model.

MichaelRoeder commented on src/main/java/org/aksw/word2vecrestful/subset/DataSubsetGenerator.java:
```diff
@@ -0,0 +1,122 @@
+ * @author Nikit
+ *
+ */
+public class DataSubsetGenerator {
```
Please document the expected file structure somewhere.

MichaelRoeder replied:
Already much better. However, could you please
* add an explanation of the single fields (it is not that much clear what "sd" means and we might even forget where the centroid came from...)
* Use a proper formatting because at the moment, javadoc sees the JSON block as text and won't care about the linebreaks, etc. that make it readable.

MichaelRoeder reviewed (changes requested):
The solution looks fine but can be improved :wink:

nikit-srivastava committed:
Implementing requested changes

1. subset files stored as a list of words separated by endline
2. subset models saved in memory at time of query
3. removing unnecessary exception declarations

nikit-srivastava committed:
documenting expected config format

nikit91 commented:
Hi @MichaelRoeder, 
I implemented the requested changes, could you please review again?

MichaelRoeder commented on src/main/java/org/aksw/word2vecrestful/subset/DataSubsetProvider.java:
```diff
@@ -0,0 +1,84 @@
+public class DataSubsetProvider {
+
+	public static String fileDir = Cfg.get("org.aksw.word2vecrestful.Application.subsetfiledir");
+	public static final Map<String, List<String>> SUBSET_MODELS = new HashMap<>();
```
I am slightly unhappy with this. Either this class itself is a singleton OR you have an instance that you are distributing via dependency injection. However, now you have made everything static which is not a good way to program it.

MichaelRoeder reviewed (changes requested):
I still would like to see some improvements :wink:

nikit-srivastava committed:
Implementing requested changes

1. Removed static handling of DataSubsetProvider
2. Better file handling for reading subset file
3. UTF8 encoding enforced on the text
4. Formatted documentation of data json

nikit-srivastava committed:
Intermediate commit

nikit-srivastava committed:
intermediate commit

nikit-srivastava committed:
intermediate commit

nikit-srivastava committed:
log4j changes

nikit-srivastava committed:
increasing jvm size

nikit-srivastava committed:
Multiplier logic fix

nikit-srivastava committed:
removing mongodb related features

nikit-srivastava committed:
Debugging messages added

nikit-srivastava committed:
changing debug to info

nikit-srivastava committed:
log correction

nikit-srivastava committed:
increasing logs

nikit-srivastava committed:
log for fetchNearbyVectors

nikit-srivastava committed:
changing twordlist to set

nikit-srivastava committed:
logic change to bitset and ids

nikit-srivastava committed:
handling null values

nikit-srivastava committed:
logic change to select subset

nikit-srivastava committed:
adding logs for normal model loading

nikit-srivastava committed:
logging minVal and maxVal

nikit-srivastava committed:
word add logic change

nikit-srivastava committed:
changing bitset iteration logic

nikit-srivastava committed:
updating 'from' and 'to' logic

nikit-srivastava committed:
logs correction

nikit-srivastava committed:
logs correction

nikit-srivastava committed:
fixing bitset logic

nikit-srivastava committed:
multplier logic fix

nikit-srivastava committed:
dimnsArr value fix

nikit-srivastava committed:
removing logs

nikit-srivastava committed:
putting time logger

nikit-srivastava committed:
adding timeLogger

nikit-srivastava committed:
adding logs to binary search

nikit-srivastava committed:
logging setting of bits

nikit-srivastava committed:
changing area values

nikit-srivastava committed:
new test scenario

nikit-srivastava committed:
setting area to 1/30th

nikit-srivastava committed:
setting area to 1/20th

nikit-srivastava committed:
sdmap logic change and new time logs

nikit-srivastava committed:
new less restrictive model added

nikit-srivastava committed:
New test created

nikit-srivastava committed:
removing unnecessary logs and putting performance identification log

nikit-srivastava committed:
updated test config and log4j properties

nikit-srivastava committed:
correcting predicted words log

nikit-srivastava committed:
adjusting config

nikit-srivastava committed:
changing multiplier type to float

nikit-srivastava committed:
configuring test vals

nikit-srivastava committed:
adjusting config vals

nikit-srivastava committed:
selecting all closest words

nikit-srivastava committed:
adjusting values

nikit-srivastava committed:
Theta model introduced

nikit-srivastava committed:
calling cosineSim for normalized vectors

MichaelRoeder committed:
Added a small class for reducing the original model to a model of a given size. Some smaller changes to reduce problems.

nikit-srivastava committed:
Model with mean comparison vector

MichaelRoeder committed:
Merge branch 'db-normal-model' of https://github.com/nikit91/Jword2vec into db-normal-model

MichaelRoeder committed:
Changed sum to be double value instead of float.

nikit-srivastava committed:
pushing test cfg changes

nikit-srivastava committed:
reducing bucket multiplier to 100 and logging bucket size

MichaelRoeder committed:
Changed the API of the getCloses methods to reduce complexity.

nikit-srivastava committed:
updating multiplier logic in theta model

nikit-srivastava committed:
fixing test logic

MichaelRoeder committed:
Added a multi-threading implementation of the IndxLR model.

MichaelRoeder committed:
Updated multithreaded solution for indexed model.

nikit-srivastava committed:
Merge branch 'db-normal-model' into changedApi

nikit-srivastava committed:
Merge pull request #2 from nikit91/changedApi

Changed api

nikit-srivastava committed:
fixing compilation error

nikit-srivastava committed:
configuring mutli threaded test

nikit-srivastava committed:
increasing k size

nikit-srivastava committed:
euclidean distance test

nikit-srivastava committed:
static method

nikit-srivastava committed:
implementing multiple vector model

nikit-srivastava committed:
fixing index logic

nikit-srivastava committed:
putting logs

nikit-srivastava committed:
null check

nikit-srivastava committed:
changing config

nikit-srivastava committed:
config change

nikit-srivastava committed:
changing config

nikit-srivastava committed:
config change

nikit-srivastava committed:
config change

nikit-srivastava committed:
config change

nikit-srivastava committed:
putting euclidean distance logic

nikit-srivastava committed:
changing nearby vec logic

nikit-srivastava committed:
making KMeans model

nikit-srivastava committed:
setting maxiterations to 10

nikit-srivastava committed:
logs included

nikit-srivastava committed:
previous model

nikit-srivastava committed:
changing config

nikit-srivastava committed:
config change

nikit-srivastava committed:
config change

nikit-srivastava committed:
comparison vecs persistence logic added

nikit-srivastava committed:
config change

nikit-srivastava committed:
adding more test cases

nikit-srivastava committed:
more centroids added to test

nikit-srivastava committed:
model change

nikit-srivastava committed:
testing with words

nikit-srivastava committed:
switching model in test

nikit-srivastava committed:
logging comparison vec size

nikit-srivastava committed:
logs added

nikit-srivastava committed:
changing config

nikit-srivastava committed:
adding logs

nikit-srivastava committed:
null check added

nikit-srivastava committed:
config change

nikit-srivastava committed:
config change

nikit-srivastava committed:
test change

nikit-srivastava committed:
removing obsolete models

nikit-srivastava committed:
refactoring the code

nikit-srivastava committed:
removing duplicate methods

nikit-srivastava committed:
removing database normalizing feature

nikit-srivastava committed:
changing access specifier

nikit-srivastava committed:
modifying access specifiers

nikit-srivastava committed:
Changing method calls

nikit-srivastava committed:
changing test config

nikit-srivastava committed:
removing unused code

nikit-srivastava committed:
deleting unused class

nikit-srivastava committed:
merged

nikit-srivastava committed:
Merge branch 'dice-group-master'

nikit-srivastava committed:
Merge pull request #4 from nikit91/db-normal-model

New search model

nikit91 commented:
New Models added:

**W2VNrmlMemModelBruteForce** : 
Brute Force model to find the closest vector to a given vector by looking for maximum cosine similarity among all vectors
**W2VNrmlMemModelBinSrch**
This class selects a given number of vectors (1 mean vector and others on basis Map iterator) and then calculates the cosine similarity of all words in model to those vectors.
**W2VNrmlMemModelKMeans**
This class selects a given number of vectors (centroids of the KMeans result on the model vectors) and then calculates the cosine similarity of all words in model to those vectors.

nikit91 commented:
@MichaelRoeder Please Review

MichaelRoeder reviewed (approved):
LGTM
